File: v1/LLD/Design-Patterns/Mediator/code/mediator.js
/**
 * Mediator Pattern — a ChatRoom routes messages between Users instead of Users talking directly.
 * Run: node mediator.js
 */

class ChatRoom {
  constructor() {
    this.users = [];
  }

  addUser(user) {
    this.users.push(user);
  }

  sendMessage(message, sender) {
    this.users.forEach(user => {
      if (user !== sender) {
        user.receive(message, sender.name);
      }
    });
  }
}

class User {
  constructor(name, mediator) {
    this.name = name;
    this.mediator = mediator;
  }

  send(message) {
    console.log(`${this.name} sends: ${message}`);
    this.mediator.sendMessage(message, this);
  }

  receive(message, fromName) {
    console.log(`${this.name} received from ${fromName}: ${message}`);
  }
}

const chatRoom = new ChatRoom();

const alice = new User("Alice", chatRoom);
const bob = new User("Bob", chatRoom);
const carol = new User("Carol", chatRoom);

chatRoom.addUser(alice);
chatRoom.addUser(bob);
chatRoom.addUser(carol);

alice.send("Hey everyone!");
bob.send("Hi Alice!");

module.exports = { ChatRoom, User };

File: v1/LLD/Design-Patterns/Mediator/code/mediator.py
"""
Mediator Pattern — a ChatRoom routes messages between Users instead of Users talking directly.
Run: python mediator.py
"""


class ChatRoom:
    def __init__(self):
        self.users = []

    def add_user(self, user):
        self.users.append(user)

    def send_message(self, message, sender):
        for user in self.users:
            if user is not sender:
                user.receive(message, sender.name)


class User:
    def __init__(self, name, mediator: ChatRoom):
        self.name = name
        self.mediator = mediator

    def send(self, message):
        print(f"{self.name} sends: {message}")
        self.mediator.send_message(message, self)

    def receive(self, message, from_name):
        print(f"{self.name} received from {from_name}: {message}")


if __name__ == "__main__":
    chat_room = ChatRoom()

    alice = User("Alice", chat_room)
    bob = User("Bob", chat_room)
    carol = User("Carol", chat_room)

    chat_room.add_user(alice)
    chat_room.add_user(bob)
    chat_room.add_user(carol)

    alice.send("Hey everyone!")
    bob.send("Hi Alice!")
